yaksh/static/yaksh/js/chat.js

```diff
@@ -0,0 +1,166 @@
+$(document).ready(function() {
+    if ($("#chat_status").val() == "True" && $("#quiz_status").val() == "False"){
+        $("#open_chat_icon").toggle();
+    }
+    var sender = $('#handle').val();
+    var chatsock;
+    var label;
+    var timer;
+    var page_title = document.title;
+    var course_id = $("#course_id").val();
+
+    $("#main").on("click", function(){
+        // Open chat window
+        var room_url = window.location.protocol + "//" +
+                    window.location.host + "/exam/new/chat/" + course_id;
+        ajax_request(room_url);
+    });
+
+    function ajax_request(chat_url) {
+        // Get messages from ajax request
+        $.ajax({
+            url: chat_url,
+            timeout:30000,
+            type: "GET",
+            dataType: 'json',
+            contentType: 'application/json; charset=utf-8',
+            success: function(msg) {
+                request_success(msg);
+            },
+            error: function(jqXHR, textStatus) {
+                alert("Unable to send messages");
+            }
+        });
+        $("#chat_container").toggle();
+    }
+
+    function request_success(msg) {
+        // On request success fill data
+        label = msg['room_label'];
+        create_ws(label);
+        $("#msg_base").empty();
+        if(msg['success']){
+            for (var j=0; j<msg['messages'].length; j++){
+                fill_data(msg['success'], msg['messages'][j]);
+            }
+        }
+        else{
+            fill_data(msg['success']);
+        }
+    }
+
+    function create_ws(label) {
+        // Create websocket connection with room
+        var ws_scheme = window.location.protocol == "https:" ? "wss" : "ws";
+        chatsock = new ReconnectingWebSocket(ws_scheme + '://' + window.location.host +
+            "/chat/" + label + "/" + course_id);
+        chatsock.onmessage = function(message) {
+            if(document.getElementById("no_msg_div") != null){
+                $("#msg_base").empty();
+            }
+            var data = JSON.parse(message.data);
+            fill_data(data['success'], data['messages'], "new")
+        };
+        chatsock.onerror = function(){
+            $("#chat_err").toggle();
+            chatsock.close();
+        }
+    }
+
+    $("#chatform").on("submit", function(event) {
+        // Submit chat message onto the websocket
+        if($('#message').text().trim().length != 0){
+            var message = {
+                sender_id: sender,
+                message: $('#message').html(),
+            }
+            chatsock.send(JSON.stringify(message));
+        }
+        $("#message").empty();
+        return false;
+    });
+
+    // Create divs for showing sent and received messages
+    var div_sent = "<div class='row msg_container base_sent'>\n"+
```

Comment:
Cant this be done with <code>document.createElement</code>? This is just horrible!